src/Symfony/Component/Form/Tests/Extension/Core/Type/CollectionTypeTest.php

```diff
@@ -315,4 +343,172 @@ public function testPrototypeSetNotRequired()
         $this->assertFalse($form->createView()->vars['required'], 'collection is not required');
         $this->assertFalse($form->createView()->vars['prototype']->vars['required'], '"prototype" should not be required');
     }
+
+    public function testEntryOptions()
+    {
+        $data = array(
+            'foo',
+            'bar',
+        );
+
+        $attr = array('class' => 'my&item&class', 'data-value' => 'my&value');
+
+        $view = $this->factory->create('Symfony\Component\Form\Extension\Core\Type\CollectionType', $data, array(
+            'allow_add' => true,
+            'prototype' => true,
+            'prototype_name' => '__test__',
+            'entry_options' => array(
+                'label' => 'Item Label:',
+                'attr' => $attr,
+            ),
+        ))->createView();
+
+        $this->assertSame('Item Label:', $view->children[0]->vars['label']);
+        $this->assertSame('Item Label:', $view->children[1]->vars['label']);
+        $this->assertSame($attr, $view->children[0]->vars['attr']);
+        $this->assertSame($attr, $view->children[1]->vars['attr']);
+        $this->assertSame('_collection_entry', $view->children[0]->vars['unique_block_prefix']);
+        $this->assertSame('_collection_entry', $view->children[1]->vars['unique_block_prefix']);
+    }
+
+    public function testEntryOptionsAsCallable()
+    {
+        $data = array(
+            'foo',
+            'bar',
+        );
+
+        $attr = array(
+            array(
+                'class' => 'foo&class',
+                'data-upper' => 'FOO',
+            ),
+            array(
+                'class' => 'bar&class',
+                'data-upper' => 'BAR',
+            ),
+        );
+
+        $view = $this->factory->create('Symfony\Component\Form\Extension\Core\Type\CollectionType', $data, array(
+            'allow_add' => true,
+            'prototype' => true,
+            'prototype_name' => '__test__',
+            'entry_options' => function ($entry) {
+                return array(
+                    'label' => ucfirst($entry).':',
+                    'attr' => array('class' => $entry.'&class', 'data-upper' => strtoupper($entry)),
+                );
+            },
+        ))->createView();
+
+        $this->assertSame('Foo:', $view->children[0]->vars['label']);
+        $this->assertSame('Bar:', $view->children[1]->vars['label']);
+        $this->assertSame($attr[0], $view->children[0]->vars['attr']);
+        $this->assertSame($attr[1], $view->children[1]->vars['attr']);
+        $this->assertSame('_collection_entry', $view->children[0]->vars['unique_block_prefix']);
+        $this->assertSame('_collection_entry', $view->children[1]->vars['unique_block_prefix']);
+    }
+
+    public function testEntryOptionsAsCallableWithPrototypeUsePrototypeOptionsData()
+    {
+        $data = array(
+            'foo',
+            'bar',
+        );
+
+        $attr = array(
```

Comment:
I would name this `$expectedAttr` and move it below the `create()` call to make the intent of the test clear. Usually, tests should be structured into the following sections (which may in part be missing):

1. Fixture setup
2. Execution of the test
3. Test verification
4. Cleanup

The `create()` statement belongs to 2, while `$attr` (`$expectedAttr`) is already part of the test verification. Hence I'd move it below.